cirq-core/cirq/study/sweeps_test.py

```diff
@@ -77,6 +77,58 @@ def test_zip():
     assert _values(sweep, 'b') == [4, 5, 6]
 
 
+def test_zip_longest():
+    sweep = cirq.ZipLongest(cirq.Points('a', [1, 2, 3]), cirq.Points('b', [4, 5, 6, 7]))
+    assert tuple(sweep.param_tuples()) == (
+        (('a', 1), ('b', 4)),
+        (('a', 2), ('b', 5)),
+        (('a', 3), ('b', 6)),
+        (('a', 3), ('b', 7)),
+    )
+    assert sweep.keys == ['a', 'b']
+    assert (
+        str(sweep) == 'ZipLongest(cirq.Points(\'a\', [1, 2, 3]), cirq.Points(\'b\', [4, 5, 6, 7]))'
+    )
+    assert (
+        repr(sweep)
+        == 'cirq_google.ZipLongest(cirq.Points(\'a\', [1, 2, 3]), cirq.Points(\'b\', [4, 5, 6, 7]))'
+    )
+
+
+def test_zip_longest_compatibility():
+    sweep = cirq.Zip(cirq.Points('a', [1, 2, 3]), cirq.Points('b', [4, 5, 6]))
+    sweep_longest = cirq.ZipLongest(cirq.Points('a', [1, 2, 3]), cirq.Points('b', [4, 5, 6]))
+    assert tuple(sweep.param_tuples()) == tuple(sweep_longest.param_tuples())
+
+    sweep = cirq.Zip(
+        (cirq.Points('a', [1, 3]) * cirq.Points('b', [2, 4])), cirq.Points('c', [4, 5, 6, 7])
+    )
+    sweep_longest = cirq.ZipLongest(
+        (cirq.Points('a', [1, 3]) * cirq.Points('b', [2, 4])), cirq.Points('c', [4, 5, 6, 7])
+    )
+    assert tuple(sweep.param_tuples()) == tuple(sweep_longest.param_tuples())
+
+
+def test_empty_zip():
+    assert len(cirq.ZipLongest()) == 0
+    with pytest.raises(ValueError, match='non-empty'):
+        _ = cirq.ZipLongest(cirq.Points('e', []), cirq.Points('a', [1, 2, 3]))
+
+
+def test_zip_eq():
+    sweep1 = cirq.ZipLongest(cirq.Points('a', [1, 2, 3]), cirq.Points('b', [4, 5, 6, 7]))
+    sweep2 = cirq.ZipLongest(cirq.Points('a', [1, 2, 3]), cirq.Points('b', [4, 5, 6, 7]))
+    sweep3 = cirq.ZipLongest(cirq.Points('a', [1, 2]), cirq.Points('b', [4, 5, 6, 7]))
+    sweep4 = cirq.Zip(cirq.Points('a', [1, 2]), cirq.Points('b', [4, 5, 6, 7]))
```

Comment:
As it is, both `sweep3 == sweep4` and `hash(sweep3) == hash(sweep4)` are `True`.
Perhaps we need to add a strict `type(other) is Zip` to the `Zip.__eq__` function
https://github.com/quantumlib/Cirq/blob/58d0205e2b90037cacc2cbf35776f2553c2e3362/cirq-core/cirq/study/sweeps.py#L294-L297

and insert `ZipLongest` or a similar type-marker to the hashed tuple at
https://github.com/quantumlib/Cirq/blob/58d0205e2b90037cacc2cbf35776f2553c2e3362/cirq-core/cirq/study/sweeps.py#L364-L365

Reply:
Changed to use 'is not Zip' and also changed the test to use Cirq's equality tester.